### PR TITLE
Drop virtual buffer attributes with invalid XML names

### LINKED ISSUE #7173: NVDA gets an error when the webpage contains aria-label syntax problem.
### Steps to reproduce:
When I open a webpage which contains aria-label syntax error, NVDA plays the error sound, and NVDA does not read the content. The syntax problem is the folloving:
&lt;section aria-label"almenük" role="region"&gt;
(= sign is missing after aria-label)
You can test it on the folloving page: https://alkosoft.hu/szolgaltatasok/tavszerviz
On the top, press d for four times, and press down arrow for three times. In this case the error sound playes, and NVDA does not read the content.

### Expected behavior:
Good solution would be that NVDA skips the aria-label, but content of the section can be navigate.

### Actual behavior:
NVDA playes an error sound, and it does not announce content of the region.

### System configuration:
NVDA version: next-14045,c75e7d19

NVDA Installed or portable: installed

Windows version: Windows 7 64 bits

Name and version of other software in use when reproducing the issue: Mozilla Firefox 53.0.2

### Other questions:

Does the issue still occur after restarting your PC?
Yes
Have you tried any other versions of NVDA?
No

### PR DESCRIPTION
**Pull request by akj**
### Link to issue number:

Fixes #7173

### Summary of the issue:

Malformed markup can make a browser expose accessibility attribute names that are not valid XML names. For example, `<section aria-label"almenük">` (note the missing `=`) yields an IA2 attribute name containing quote characters. NVDA stores virtual buffer fields as XML markup, so emitting such a name makes expat reject the buffer content: NVDA plays an error sound and cannot read the affected range in browse mode.

Minimal test case (from @jcsteh), pasted into the Firefox address bar:

```
data:text/html,<section aria-label"almen%C3%BCk" role="region">test
```

### Description of user facing changes:

Browse mode now reads pages containing this kind of malformed markup normally; the broken attribute is ignored.

### Description of developer facing changes:

`VBufStorage_fieldNode_t::addAttribute` now rejects attributes whose names are not valid XML names, returning `false`. This applies to all virtual buffer backends.

### Description of development approach:

Validation happens once, in `addAttribute`, the attribute map's single insertion point, so the buffer only ever holds names that are valid in the XML it serializes to. `isValidXMLAttribName` in `nvdaHelper/common/xml.h` accepts a conservative ASCII subset of XML names: the expat build Python ships enforces the stricter XML 1.0 fourth edition name rules, and every attribute name NVDA's Python code consumes is ASCII, so anything outside that subset can only be junk from broken markup. Because only valid names can reach the map, `generateAttributesForMarkupOpeningTag` now writes stored names directly and `sanitizeXMLAttribName` is deleted.

Dropping the attribute extends the tradeoff made in #6249, which rewrote spaces in names to underscores: a readable buffer is worth more than preserving an attribute nobody can parse. The practical difference is that a name with invalid characters, such as a localised action name containing spaces, is now absent from the buffer instead of exposed under a mangled name. The well-known ASCII names NVDA itself consumes are unaffected.

An existing XML library was considered and rejected; see the review discussion below for the survey.

### Testing strategy:

- GitHub Actions for the head commit: build, lint, type checks, unit tests, and all but two system test jobs pass.
- The two red jobs (`chrome_misc_aria` on windows-2025-vs2026, `l10n` on windows-2022) each pass on the other OS in the same run, and master is currently red on two different system test suites, so both look like known runner flakiness rather than this change.
- Manual: loaded the Firefox test case above; browse mode reads "test" with no error sound.

### Known issues with pull request:

None in the changed code. Two required system test jobs are red; see testing strategy.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

Notes on the checklist: a change log entry is included; no user or developer documentation or GUI help changes are needed. There is no unit test harness for the `vbufBase` C++ code; the change is a simple rejection at the buffer's insertion point, covered by the manual test case. Add-on facing behaviour changes only for attribute names that were never valid XML: they are now dropped instead of exposed under a sanitized name.

### CODE REVIEW

**Comment by SaschaCowley** [comment]
@akj have you investigated using an existing XML sanitization library to do this?
* If so, which ones, what were the results, and why did you go with this approach?
* If not, why not?
I'm a little concerned about this kind of bespoke logic, especially for future maintainers.

**Comment by akj** [comment]
Yes, I did look into this before settling on the current approach — and I share the concern about bespoke logic; it's a big part of why the fix ended up shaped the way it is.

* **What I investigated:** pugixml, TinyXML-2, XmlLite, and libxml2 on the C++ side, and lenient parsing with lxml on the Python side.
* **The results:** none of them solve this particular problem. XML libraries are good at escaping attribute *values* (which the buffer already does), but their writers all assume the caller supplies valid attribute *names* — none validate or repair them. The closest thing is libxml2's `xmlValidateName`, but that only tells you a name is bad; the repair logic would still be ours to write, and it would mean adding a C library to the code we inject into other processes. So a library wouldn't remove the bespoke part — it would just sit next to it.
* **Why this approach:** there's also a subtlety that made me warier of a library: "valid" depends on the parser doing the reading, and Python's expat follows the older, stricter XML name rules — a modern spec-conformant validator passes names our own parser then rejects. The safest definition of valid here is "whatever NVDA's parser accepts", and no library encodes that.

On the maintainer burden specifically: rather than asking future maintainers to trust a character list in a C++ header, I've added unit tests (`tests/unit/test_XMLFormatting.py`) asserting that every name the sanitizer can produce is accepted by the same expat parser NVDA uses, plus the specific problem cases from this issue and #6249. If expat's rules ever shift, CI will say so rather than a user hitting it. For scale, the sanitizer is about 20 lines, in the same header where `appendCharToXML` has been doing the equivalent bespoke job for attribute values since 2016.

While revisiting this I also tightened the fix based on further review: names are now sanitized once when an attribute is added to the buffer rather than on every serialization, and a couple of edge cases (invalid leading characters, name collisions after sanitization) are handled.

File: tests/unit/test_XMLFormatting.py
Comment (SaschaCowley) [inline tests/unit/test_XMLFormatting.py:1]: Most of these tests are not useful, since they aren't actually testing the code that gets exercised by NVDA. Ultimately, most of this should be covered by system tests, but, as we don't currently have the infrastructure for testing against Firefox, I would suggest that most of these should be removed.

File: nvdaHelper/common/xml.h [hunk 1]
```diff
@@ -38,10 +38,36 @@ inline void appendCharToXML(const wchar_t c, std::wstring& xml, bool isAttribute
...
+	// Python enforces the stricter XML 1.0 fourth edition name rules, so anything
+	// broader risks emitting names it rejects. All attribute names NVDA consumes
+	// are ASCII, so nothing meaningful is lost by replacing the rest.
+	return (c >= L'a' && c <= L'z')
+		|| (c >= L'A' && c <= L'Z')
+		|| (c >= L'0' && c <= L'9')
+		|| c == L'-' || c == L'.' || c == L':' || c == L'_';
+}
+
 inline std::wstring sanitizeXMLAttribName(std::wstring attribName) {
```
Comment (SaschaCowley) [inline nvdaHelper/common/xml.h]: @jcsteh, what was the logic for sanitizing invalid attribute names, rather than just dropping them? The only reason I can think to do it is that production builds don't emit C++ logs, so keeping them provides some level of visibility into the original XML.

File: user_docs/en/changes.md [hunk 2]
```diff
@@ -48,6 +48,7 @@
 
 ### Bug Fixes
 
+* In browse mode, NVDA no longer plays an error sound and fails to read content when a page contains malformed markup such as `aria-label"foo"`, which produced accessibility attribute names containing characters that are invalid in XML. (#7173, @akj)
```
Comment (SaschaCowley) [inline user_docs/en/changes.md]:
```suggestion
* In browse mode in Mozilla Firefox, NVDA no longer fails to read content containing markup with invalid XML attribute names. (#7173, @akj)
```

**Review by SaschaCowley — CHANGES_REQUESTED** [review]

File: nvdaHelper/common/xml.h (hunk 1 above)
Comment (jcsteh) [reply nvdaHelper/common/xml.h]: Off the top of my head, I don't recall. I can look into it if you'd like, but I'd basically just be doing the same git blame/log/GitHub issue trawling that you would. That said, my recall for such things is usually fairly good, so the fact that I don't remember anything significant is maybe possibly an indicator that it wasn't particularly important? 🤷‍♂️

If you do chase this down and end up in an issue that doesn't make sufficient sense, I'm happy to take a look; it might jog my memory.

Comment (jcsteh) [reply nvdaHelper/common/xml.h]: Oh blah, sorry, I see my PR and commit is linked from the top of this PR, no blame necessary. Yeah, I'd say you're exactly right: I was probably just trying to make sure we could debug this. If stripping them reduces complexity, maybe we should just do that, though I'm not sure how much of a difference it really makes; you still have to detect the invalid names in the first place.

Comment (SaschaCowley) [reply nvdaHelper/common/xml.h]: @jcsteh I was thinking that:
* A function that reports whether an attribute name is valid can return on the first invalid character. The complexity is still O(n), but there's an early escape rather than continuing to sanitize the name, which renders it useless anyway.
* The current logic here involves prepending a char to the string if it begins with an invalid character, which is `O(n)`. Fixing that requires either changing the logic here (not terribly difficult, but very slightly messy, at least how it looks in my head), or just dropping invalid attribute names. Though come to think of it, many of the code changes required to nicely do the detection would be the same for a more optimal sanitization approach...
* Dropping invalid attribute names means they're not taking up space, which is unlikely to make a real-world difference, but doesn't hurt if they aren't doing anything.

I'm not fussed either way, but given we can't really use the sanitized names for anything but some fairly restricted cases of debugging, I wonder what the best approach is.

Comment (jcsteh) [reply nvdaHelper/common/xml.h]: Fair. Impossible to say whether this makes that much of a difference in practice without profiling it, but also, if it isn't useful, there's not a lot of point in keeping it. Dropping them seems reasonable. Browser DevTools have come a long way since this was done, so it's much easier to debug problems like this by looking at the problematic element anyway.

**Comment by SaschaCowley** [comment]
@akj is this ready for re-review?

**Comment by akj** [comment]
@SaschaCowley Thank you for your patience. I've changed the PR to reflect the discussion in this thread, and updated the title and description to match the new approach. This is now ready for re-review.

File: user_docs/en/changes.md [hunk 3]
```diff
@@ -64,6 +64,7 @@
 ### Bug Fixes
 
 * 64-bit NVDA now reports the correct location and label for Win32 menu items of 32-bit applications when display scaling is above 100%. (#19225, #20158, @christopherpross)
+* In browse mode in Mozilla Firefox, NVDA no longer fails to read content containing markup with invalid XML attribute names. (#7173, @akj)
```
Comment (seanbudd) [inline user_docs/en/changes.md]: Please move the changelog entry to 2027.1

**Review by SaschaCowley — APPROVED** [review]
Thanks, @akj
